lib/spack/spack/test/llnl/util/tty/log.py

```diff
@@ -0,0 +1,472 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from __future__ import print_function
+import contextlib
+import os
+import re
+import signal
+import time
+
+try:
+    import termios
+except ImportError:
+    termios = None
+
+import pytest
+
+import llnl.util.tty.log
+from llnl.util.tty.log import log_output
+from llnl.util.tty.pty import PseudoShell
+
+from spack.util.executable import which
+
+
+def test_log_python_output_with_python_stream(capsys, tmpdir):
+    # pytest's DontReadFromInput object does not like what we do here, so
```

Comment:
I think that's a follow-on PR thing.

Reply:
This module has been entirely recreated from scratch: IMO now is the best time to consider whether to do it.

Reply:
I’d rather not do this refactor. I don’t know why we should implement this outside pytest when it’s intended to be there for testing.

Reply:
> Does capsys.disabled() effectively undo all modifications done by PyTest to sys.stdout, sys.stdin etc.?

Based on looking at the source code of `_pytest/capture` (and specifically the `FDCapture` and `SysCapture` objects, I think the answer to this is "yes".

That leaves the second question. Like I mention above, I think that issue can be resolved by using `capsys.disabled` for all tests in this module.

Reply:
These are refactored and no longer use `capsys`.

Reply:
My overall concern is that our fd management and PyTest's fd management (whether it be capsys or capfd) will create an environment that doesn't match where the code actually runs. Overall from reading `FDCapture` I think if it were disabled with `capfd.disabled()` then it wouldn't be an issue. I don't think it's an issue for tests using `PseudoShell` (since it explicitly configures the FD's for the child to inherit).